File: 03-creational-patterns/03-builder/03-query-builder.go
package main

import (
	"fmt"
	"strings"
)

type QueryBuilder struct {
	selectCols []string
	table      string
	where      string
	orderBy    string
}

func NewQueryBuilder() *QueryBuilder {
	return &QueryBuilder{}
}

func (qb *QueryBuilder) Select(cols ...string) *QueryBuilder {
	qb.selectCols = cols
	return qb
}

func (qb *QueryBuilder) From(table string) *QueryBuilder {
	qb.table = table
	return qb
}

func (qb *QueryBuilder) Where(condition string) *QueryBuilder {
	qb.where = condition
	return qb
}

func (qb *QueryBuilder) OrderBy(order string) *QueryBuilder {
	qb.orderBy = order
	return qb
}

func (qb *QueryBuilder) Build() string {
	var sb strings.Builder
	sb.WriteString("SELECT ")
	if len(qb.selectCols) > 0 {
		sb.WriteString(strings.Join(qb.selectCols, ", "))
	} else {
		sb.WriteString("*")
	}
	sb.WriteString(" FROM ")
	sb.WriteString(qb.table)
	if qb.where != "" {
		sb.WriteString(" WHERE ")
		sb.WriteString(qb.where)
	}
	if qb.orderBy != "" {
		sb.WriteString(" ORDER BY ")
		sb.WriteString(qb.orderBy)
	}
	sb.WriteString(";")
	return sb.String()
}

func main() {
	query := NewQueryBuilder().
		Select("id", "name").
		From("users").
		Where("active = 1").
		OrderBy("name").
		Build()

	fmt.Println(query)
}

File: 03-creational-patterns/03-builder/03-query-builder.py
class QueryBuilder:
    def __init__(self):
        self._select = []
        self._from = ""
        self._where = ""
        self._order_by = ""

    def select(self, *cols):
        self._select = cols
        return self

    def from_table(self, table):
        self._from = table
        return self

    def where(self, condition):
        self._where = condition
        return self

    def order_by(self, column):
        self._order_by = column
        return self

    def build(self):
        query = "SELECT "
        query += ", ".join(self._select) if self._select else "*"
        query += f" FROM {self._from}"
        if self._where:
            query += f" WHERE {self._where}"
        if self._order_by:
            query += f" ORDER BY {self._order_by}"
        return query + ";"

# Client
if __name__ == "__main__":
    query = (
        QueryBuilder()
        .select("id", "name")
        .from_table("users")
        .where("active = 1")
        .order_by("name")
        .build()
    )

    print(query)
